### ISSUE TITLE
 Reading Frame position failed when reading a volume

### ISSUE THREAD

**Issue by arngaillard** [openslide/openslide#484]
### Operating system

macOS

### Platform

64-bit ARM

### OpenSlide version

latest on develop

### Slide format

DICOM

### Issue details

Hello,

I am building a web server on top of openslide to access thumbnails/tiles of WSI slides. For a specific DICOM slide (that I do not know yet if I can share), I receive the following error

```
libdicom Parse error: Reading Frame position failed - Row must be less than 3
```

I tried to open this volume with a python library [wsidicom](https://github.com/imi-bigpicture/wsidicom) and the retrieval of the tiles at the faulty levels or the thumbnail doesn't crash. This library seems to use `gdcm` to open DICOM files. My question is, is this an error on `libdicom`/`openslide` side or is it rather than `gdcm`/`wsidicom` doesn't check this error ? 

I am sorry I cannot provide the slide for the moment, I'll ask and see if this is possible to share.

Also, this isn't the `libdicom` issue tracker but there seems to be some common developers between to two and I'm not using the lib directly but openslide. Please tell me if this issue belongs in `libdicom` instead.

**Comment by bgilbert** [comment]
Thanks for reporting this, and for testing the new DICOM support!  That error comes from libdicom, but let's see what @jcupitt thinks.

**Comment by arngaillard** [comment]
Hi @jcupitt  and @bgilbert,

Thank you for the reactivity.

> libdicom is probably confused and your slide is OK, so it'd be very helpful to get some sample data. If you could possibly send the slide (perhaps to my private email? [jcupitt@gmail.com](mailto:jcupitt@gmail.com)) that'd be fantastic.
> 
The owners of the slide aren't very keen to share it. Even sending you the `dcm-dump` all of the files seems to be an issue. I'm discussing with them at the moment to see if the dump could be sent if all the PHI is manually stripped from it.

> And the exact command you ran to trigger the error, of course.
> 
The error occurs when I try to obtain a thumbnail (through `openslide-python`'s method `get_thumbnail`). It looks like it calls the C function `openslide_read_region` right before failing  (with parameters `x=0`, `y=0`, `level=8`, `h=364` and `w=817 `). Though these aren't probably the most useful without any data...

> I guess you're using my libdicom, is that right? The main branch of https://github.com/jcupitt/libdicom

Yes, this is the version I have installed (as found in the `subprojects` directory of the openslide repo).

**Comment by jcupitt** [comment]
Do you know what library / scanner wrote the DICOM? I might be able to find a similar file in the NEMA archive.

**Comment by arngaillard** [comment]
The `Manufacturer` tag says `3DHISTECH Kft`, however, `ManufacturerModelName` is empty.

**Comment by jcupitt** [comment]
Hmm, my 3dhistech samples work OK :( I think I'll need more information.

**Comment by bgilbert** [comment]
@arngaillard If the issue is reproducible with particular scan settings, the owners could also scan some random slide that's okay to share.

**Comment by arngaillard** [comment]
I have sent @jcupitt the dcm-dumps of the slides, hoping that this will be of some help.
Unfortunately I cannot get in touch with the owners of the scanner/environnement that has been used to generate those slides for the moment.

**Comment by jcupitt** [comment]
Quick update: I've fixed a few issues in libdicom, and I think I can see how to patch openslide to fix this.

Unfortunately, thanks to some pending renames, the libdicom that openslide needs has become out of step with libdicom master :( I think we should probably get libdicom 1.0 agreed before patching DICOM support in openslide.

**Comment by jcupitt** [comment]
It looks like there are two remaining issues:

1. libdicom was using `DimensionIndex` to get the tile position from `PerFrameFunctionalGroupSequence`, but your file isn't using that, it just saves the tile positions in pixels in `ColumnPositionInTotalImagePixelMatrix`
2. Your file isn't setting `DimensionOrganizationType` to sparse, but it does have missing tiles.

So we need changes to libdicom and to openslide:

1. libdicom needs to use the tile (x, y) position rather than the index, it needs to flip to parse mode automatically if there's a valid `PerFrameFunctionalGroupSequence`, and it needs a way to report missing tiles
2. `openslide-vendor-dicom.c` needs to handle missing tiles from libdicom

I have a couple of PRs ready, I'll make them now.

**Comment by jcupitt** [comment]
Here's the openslide PR which (I think) fixes this.

https://github.com/openslide/openslide/pull/501

**Comment by bgilbert** [comment]
#501 has now landed.  @arngaillard, could you try again with current Git `main`?

**Comment by bgilbert** [comment]
Okay, thanks for checking.  Is it the same error message?

**Comment by arngaillard** [comment]
Yes, still `libdicom Parse error: Reading Frame position failed - Row must be less than 3`

**Comment by jcupitt** [comment]
Oh dear. OK, I'll try again.

**Comment by jcupitt** [comment]
I think I found another problem. Any chance you could test that PR, @arngaillard ?

Thank you again for working on this.

**Comment by arngaillard** [comment]
Amazing, the file no longer produces an error and I'm able to fetch tiles & the whole thumbnail.
Thank you for the reactivity !

**Comment by jcupitt** [comment]
Success! I'll merge that change, and thanks again for raising this issue and testing the PRs.
